File: apps/myjobhunter/backend/app/schemas/application/application_event_update_request.py
"""Pydantic schema for PATCH /applications/{id}/events/{event_id} body.

Allowlists only the two user-input columns on ``ApplicationEvent``:
``interview_details`` and ``note``.  Every other column
(``event_type``, ``occurred_at``, ``source``, ``email_message_id``,
``raw_payload``) is structurally immutable — the audit invariant
"this event was logged at time X via source Y" stays trustworthy.

``extra='forbid'`` rejects any attempt to PATCH the immutable fields.

The service layer enforces the additional rule that PATCH is only
valid for ``interview_scheduled`` / ``interview_completed`` event
types — system-generated events (``applied`` from kanban transitions,
``email_received`` from Gmail sync) remain immutable.

Field-level validation for ``interview_details`` mirrors
``InterviewDetailsRequest`` exactly so the create and edit paths
accept the same shape.
"""
from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.application.application_event_create_request import (
    InterviewDetailsRequest,
)

_NOTE_MAX_LEN = 5000


class ApplicationEventUpdateRequest(BaseModel):
    """Body for PATCH /applications/{application_id}/events/{event_id}.

    Both fields are optional — a caller updating only the note keeps
    ``interview_details`` unchanged on the row, and vice versa. Both
    fields are nullable: passing ``null`` clears the column.
    """

    interview_details: InterviewDetailsRequest | None = None
    note: str | None = Field(default=None, max_length=_NOTE_MAX_LEN)

    model_config = ConfigDict(extra="forbid")

File: apps/myjobhunter/frontend/src/types/application-event-update-request.ts
import type { InterviewDetails } from "./interview-details";

/**
 * Body for PATCH /applications/{id}/events/{event_id}. Mirrors
 * `ApplicationEventUpdateRequest` in
 * apps/myjobhunter/backend/app/schemas/application/application_event_update_request.py.
 *
 * Only the two user-input columns are editable; every other event
 * field is structurally immutable (backend rejects extras with 422).
 *
 * Field omission semantics: omitting a key leaves the column untouched;
 * sending `null` clears the column. Mirrors `exclude_unset=True` on
 * the backend Pydantic model.
 *
 * Backend additionally enforces that the targeted event's event_type
 * is `interview_scheduled` or `interview_completed` — non-interview
 * events return 422 with detail `event_type does not support editing`.
 */
export interface ApplicationEventUpdateRequest {
  interview_details?: InterviewDetails | null;
  note?: string | null;
}
